### PR TITLE
KOGITO-514 added empty state and loading spinners

### PR DESCRIPTION
**Pull request by AjayJagan**
- Added loading spinners when data is getting fetched(simulated with a setTimeout for testing).
- Added empty state whenever the no data is available.

### CODE REVIEW

**Comment by AjayJagan** [comment]
Hi @cristianonicolai , can you please review this as well :)

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 1]
```diff
@@ -148,8 +153,10 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
         },
         fetchPolicy:'network-only'
       });
-      setchildList(data['data']);
-      setisLoaded(true);
+      setTimeout(() => {
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: This timeout is so we can view the spinner while loading right? Wouldnt this be only applicable in the mock server, we shouldn't delay this for the actual execution.

File: packages/management-console/src/components/Templates/DataListComponent/DataListComponent.tsx [hunk 2]
```diff
@@ -26,6 +28,7 @@ const DataListComponent: React.FC<{}> = () => {
       const [isSuspendedChecked, setisSuspendedChecked] = useState<boolean>(false);
   const [initData, setInitData] = useState<any>([]);
   const [checkedArray, setCheckedArray] = useState<any>(['ACTIVE']);
+  const [isLoading, setisLoading] = useState(true);
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Templates/DataListComponent/DataListComponent.tsx]:
```suggestion
  const [isLoading, setIsLoading] = useState(true);
```

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 3]
```diff
@@ -148,8 +153,10 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+        setchildList(data['data']);
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
        setChildList(data['data']);
```

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  useEffect(() => {
+    setInitData(data);
+  }, [data]);
+
+  if (loading || isLoading) {
+    return (
+      <Bullseye>
+        <Spinner /> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
+        <TextContent>
+          <Text component={TextVariants.h2}>Loading Instances ...</Text>
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: perhaps just Loading...

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    fetchPolicy: 'network-only'
+  });
+
+  useEffect(() => {
+    setInitData(data);
+  }, [data]);
+
+  if (loading || isLoading) {
+    return (
+      <Bullseye>
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx:69]: I think this is a bit different from the suggest components in PatternFly, see: https://www.patternfly.org/v4/documentation/react/components/emptystate#spinner

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 6]
```diff
@@ -280,15 +387,25 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
                   state={child.state}
                   managementEnabled={child.managementEnabled}
-                  error = {child.error}
+                  error={child.error}
                   endpoint={child.endpoint}
                 />
               );
             })}
+          {!isLoaded && (
+            <Bullseye>
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx:405]: same comment here, seems to implement using diiferent components from PatternFly https://www.patternfly.org/v4/documentation/react/components/emptystate#spinner

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 7]
```diff
@@ -280,15 +387,25 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+              <Spinner /> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
+              <TextContent>
+                <Text component={TextVariants.h2}>
+                  Loading Child Instances ...
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: Would suggest using just Loading...

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  return (
+    <DataList aria-label="Expandable data list example">
+      {!loading &&
+        initData !== undefined &&
+        initData.ProcessInstances.map((item, index) => {
+          return (
+            <DataListItemComponent
+              id={index}
+              key={item.id}
+              instanceState={item.state}
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: cant we use a ProcessInstance object type in here instead of passing single attributes?

File: packages/management-console/src/components/Templates/DashboardComponent/Dashboard.tsx [hunk 9]
```diff
@@ -1,25 +1,39 @@
...
-            <Page header={
-                <HeaderComponent/>} skipToContent={PageSkipToContent} mainContainerId={pageId} className="page">
-                <Route exact path="/" render={() => <Redirect to="/ProcessInstances"/>}/>
-                <Route exact path="/ProcessInstances" component={DataListComponent}/>
-                <Route exact path="/ProcessInstances/:instanceID" component={ProcessDetailsPage}/>
-            </Page>
-        </React.Fragment>
-    );
+  return (
+    <React.Fragment>
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Templates/DashboardComponent/Dashboard.tsx]: Is this done by auto format? To be honest, the previous code, at least for me, looks better to follow

**Review by cristianonicolai — CHANGES_REQUESTED** [review]
@AjayJagan its a bit hard to follow what exactly changed in the PR as there are many changes regarding formatting... is that something that could be avoided?

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx (hunk 4 above)
Comment (AjayJagan) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: @cristianonicolai - Prettier is applied only when a commit is made and these files are being commited for the first time after the prettier config is added , hence we see the changes. From next time there would be not problem :100:

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx (hunk 5 above)
Comment (AjayJagan) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx:69]: @srambach - Can you please suggest the spinner to use ..

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 7 above)
Comment (srambach) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: If what you are loading is always child instances, I would say it's actually good to be specific. However, I'd use sentence case and no space before the ellipsis. "Loading child instances..."

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx (hunk 4 above)
Comment (srambach) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: Here also, if it's definitely instances it is loading, then I'd still say "Loading instances..."

File: packages/management-console/src/components/Templates/DataListContainer/DataListContainer.tsx [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+  return (
+    <React.Fragment>
+      <PageSection variant="light">
+        <DataListTitleComponent />
+        <Breadcrumb>
+          <BreadcrumbItem>
+            <Link to={'/'}>Home</Link>
+          </BreadcrumbItem>
+          <BreadcrumbItem isActive>Process Instances</BreadcrumbItem>
```
Comment (srambach) [inline packages/management-console/src/components/Templates/DataListContainer/DataListContainer.tsx]: We should use sentence case (the first letter of only the first word capitalized) so this would be "Process instances"

File: packages/management-console/src/components/Atoms/EmptyStateComponent/EmptyStateComponent.tsx [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+import { SearchIcon } from '@patternfly/react-icons';
+
+interface IOwnProps {}
+const EmptyStateComponent: React.FC<IOwnProps> = () => {
+  return (
+    <Bullseye>
+      <EmptyState>
+        <EmptyStateIcon icon={SearchIcon} />
+        <Title size="lg">No results found</Title>
+        <EmptyStateBody>Please try using different filters</EmptyStateBody>
```
Comment (srambach) [inline packages/management-console/src/components/Atoms/EmptyStateComponent/EmptyStateComponent.tsx]: Change to "Try using different filters" - guidance is not to include "please" even though we are very polite :-)

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 7 above)
Comment (cristianonicolai) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: perhaps can we keep _Loading process instances..._ ? Same as the main loading one...

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx (hunk 4 above)
Comment (cristianonicolai) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: in that case, I think would be better to use _Loading process instances..._

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx (hunk 5 above)
Comment (srambach) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx:69]: After looking at it, I agree - I think the component to use here is the one @cristianonicolai links to. So, it's the empty state component but the loading version of it. Here's the empty state table example https://patternfly-react.surge.sh/patternfly-4/documentation/react/components/table#empty-state

File: packages/management-console/src/components/Atoms/EmptyStateComponent/EmptyStateComponent.tsx (hunk 11 above)
Comment (AjayJagan) [reply packages/management-console/src/components/Atoms/EmptyStateComponent/EmptyStateComponent.tsx]: sure :)

File: packages/management-console/src/components/Templates/DataListContainer/DataListContainer.tsx (hunk 10 above)
Comment (AjayJagan) [reply packages/management-console/src/components/Templates/DataListContainer/DataListContainer.tsx]: sure thing :+1:

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx (hunk 8 above)
Comment (AjayJagan) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: done :+1:

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 7 above)
Comment (AjayJagan) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: added Loading process instances...

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 6 above)
Comment (AjayJagan) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx:405]: added the spinner from Emptystate.

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx (hunk 4 above)
Comment (AjayJagan) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: added Loading process instances...

File: packages/management-console/src/components/Atoms/SpinnerComponent/SpinnerComponent.tsx [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  const Spinner = () => (
+    <span
+      className="pf-c-spinner"
+      role="progressbar"
+      aria-valuetext="Loading..."
+    >
+      <span className="pf-c-spinner__clipper" />
+      <span className="pf-c-spinner__lead-ball" />
+      <span className="pf-c-spinner__tail-ball" />
+    </span>
```
Comment (AjayJagan) [inline packages/management-console/src/components/Atoms/SpinnerComponent/SpinnerComponent.tsx]: @srambach - 'pf-c-spinner' and its sub classes are not getting loaded. I have tested the same in sandbox provided by PF4 and the spinner is not working there as well. Can you please review this part of code and help me understand the problem :)

Comment (AjayJagan) [reply packages/management-console/src/components/Atoms/SpinnerComponent/SpinnerComponent.tsx]: here is a link for the sandbox for Spinners - https://codesandbox.io/s/odmfn

Comment (srambach) [reply packages/management-console/src/components/Atoms/SpinnerComponent/SpinnerComponent.tsx]: @AjayJagan I think I have an answer - the way they are doing it on the PF site is because it's experimental and for now it must be like that there. But, we should import this:
`import { Spinner } from '@patternfly/react-core/dist/esm/experimental';`

and then use this for the icon:
`      <EmptyStateIcon variant="container" component={Spinner} />`

That should pull in the CSS needed for the experimental component and put it in the right place in the empty state component.

Comment (srambach) [reply packages/management-console/src/components/Atoms/SpinnerComponent/SpinnerComponent.tsx]: (and then you can remove all the stuff in `const Spinner`)

Comment (AjayJagan) [reply packages/management-console/src/components/Atoms/SpinnerComponent/SpinnerComponent.tsx]: done :+1:  it works :v:

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  }, [data]);
+
+  if (loading || isLoading) {
+    return (
+      <Bullseye>
+        <SpinnerComponent spinnerText="Loading process instances..." />
+      </Bullseye>
+    );
+  }
+  if (error) return <p>oops ... error while loading</p>;
```
Comment (srambach) [inline packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: Is it in the scope of this PR to fix this? It should likely be an error alert rather than just text, but I'm not sure which kind of alert (possibly inline?)
image

Comment (AjayJagan) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: @srambach - I can add it if required. 
https://www.patternfly.org/v4/documentation/react/components/alert - is Danger Alert the one to be used or should it be and info?

Comment (AjayJagan) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: errorwhileloading
is this fine

Comment (srambach) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: 3 things :-) 
- Let's use an inline alert variation
- Can it replace the whole table? In other words, don't show the toolbar at all if it doesn't load
- Is there an action that the user can take - like reloading? If so, that can be added as the alert action and it will show up on the right. Maybe something like
`<Alert variant="danger" isInline title="Error loading process instances" action={<AlertActionLink>Retry</AlertActionLink>} />`
image

Comment (srambach) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: Ah.. I just saw your update to make the empty state look like an error state. Let me get some design feedback on that. Liz and Kyle have been in the F2F all day today so I haven't been able to catch them. Brian gave me input on the inline alert from the previous comment. ^^^

Comment (srambach) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: OK, sorry - forget all that! 
@bdellasc says:
OK, I think that works - with a few changes:
- use the Font Awesome 5 Free "exclamation-triangle" for the icon (it's a solid triangle with the exclamation point) - we're moving away from the outline status icons
- Use "Warning alert color: --pf-global--warning-color--100" for the icon
- Give a little more space between the bottom of the empty state and the bottom edge of the empty table.

I can give you information on adjusting the margin and changing the color using utilities/modifiers but wanted to pass this along asap since you'll be back to work before I am :)

Comment (AjayJagan) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: errorwhileloading

I have used font awsome icons,  --pf-global--warning-color--100 and the spacing has been added. I have pushed the code. Please review whenever time permits :)

**Review by cristianonicolai — DISMISSED** [review]

File: packages/management-console/src/components/Atoms/EmptyStateComponent/EmptyStateComponent.css [hunk 14]
```diff
@@ -0,0 +1,3 @@
+.font-style {
+  padding-bottom: 10px;
```
Comment (srambach) [inline packages/management-console/src/components/Atoms/EmptyStateComponent/EmptyStateComponent.css]: Let's take this out and use the Patternfly utility classes instead.

File: packages/management-console/src/components/Atoms/EmptyStateComponent/EmptyStateComponent.tsx [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+  EmptyState,
+  EmptyStateIcon,
+  EmptyStateBody,
+  Bullseye,
+  Button
+} from '@patternfly/react-core';
+import { SearchIcon } from '@patternfly/react-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faExclamationTriangle } from '@fortawesome/free-solid-svg-icons';
+import './EmptyStateComponent.css';
```
Comment (srambach) [inline packages/management-console/src/components/Atoms/EmptyStateComponent/EmptyStateComponent.tsx]: Let's replace this with `import '@patternfly/patternfly/patternfly-addons.css';`

File: packages/management-console/src/components/Atoms/EmptyStateComponent/EmptyStateComponent.tsx [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+  return (
+    <Bullseye>
+      <EmptyState>
+        {iconType === 'searchIcon' && <EmptyStateIcon icon={SearchIcon} />}
+        {iconType === 'warningTriangleIcon' && (
+          <FontAwesomeIcon
+            icon={faExclamationTriangle}
+            size="3x"
+            color="var(--pf-global--warning-color--100)"
+            className="font-style"
```
Comment (srambach) [inline packages/management-console/src/components/Atoms/EmptyStateComponent/EmptyStateComponent.tsx]: Here we can use  `className="pf-u-mb-xl"`

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.css [hunk 17]
```diff
@@ -0,0 +1,4 @@
+.emptystate-style {
+  padding-bottom: 20px;
+  padding-top: 20px;
+}
```
Comment (srambach) [inline packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.css]: Let's remove this and use the Patternfly utility spacing class instead.

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx [hunk 18]
```diff
@@ -0,0 +1,113 @@
+import React, { useEffect } from 'react';
+import { DataList, Bullseye } from '@patternfly/react-core';
+import DataListItemComponent from '../../Molecules/DataListItemComponent/DataListItemComponent';
+import gql from 'graphql-tag';
+import { useQuery } from '@apollo/react-hooks';
+import SpinnerComponent from '../../Atoms/SpinnerComponent/SpinnerComponent';
+import EmptyStateComponent from '../../Atoms/EmptyStateComponent/EmptyStateComponent';
+import './DataListComponent.css';
```
Comment (srambach) [inline packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: This can be `import '@patternfly/patternfly/patternfly-addons.css';` instead

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+    return (
+      <Bullseye>
+        <SpinnerComponent spinnerText="Loading process instances..." />
+      </Bullseye>
+    );
+  }
+  if (error) {
+    setIsError(true);
+    return (
+      <div className="emptystate-style">
```
Comment (srambach) [inline packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: Then we can remove this style and use the PatternFly utility `.pf-u-my-xl` instead (utility class, margin in the y direction, xl size)

**Comment by srambach** [comment]
bitmoji

**Review by srambach — APPROVED** [review]

**Review by Sara4994 — APPROVED** [review]

**Review by cristianonicolai — APPROVED** [review]
